qp_spades/tests/test_spades.py

```diff
@@ -80,46 +83,147 @@ def _generate_testing_files(self):
+        with open(file_list_fp) as fp:
+            obs_file_list_fp = fp.readlines()
+        params['out_dir'] = out_dir
+        params['environment'] = self.environment
+        self.assertEqual(''.join(obs_main_qsub_fp), EXP_MAIN.format(**params))
+        self.assertEqual(
+            ''.join(obs_finish_qsub_fp), EXP_FINISH.format(**params))
+        self.assertEqual(''.join(obs_file_list_fp),
+                         EXP_FILE_LIST.format(directory=directory))
 
-        success, ainfo, msg = spades(self.qclient, jid, params, out_dir)
-        self.assertEqual(msg[:20], 'Error running spades')
-        self.assertFalse(success)
-
-        # testing meta/no-merge
+        # testing isolate/merge
+        # note that we don't need to recreate all the above variables and we
+        # can reuse
         params = {
             'type': 'meta', 'merging': 'flash 65%', 'input': self.aid,
             'threads': 5, 'memory': 200, 'k-mers': '21,33,55,77,99,127'}
-        data = {'user': 'demo@microbio.me',
-                'command': dumps(['qp-spades', '2021.05', 'spades v3.15.2']),
-                'status': 'running',
-                'parameters': dumps(params)}
-        jid = self.qclient.post('/apitest/processing_job/', data=data)['job']
-        out_dir = mkdtemp()
-        self._clean_up_files.append(out_dir)
-
-        success, ainfo, msg = spades(self.qclient, jid, params, out_dir)
-        self.assertEqual(msg, '')
-        self.assertTrue(success)
-        self.assertEqual(ainfo[0].artifact_type, 'FASTA_preprocessed')
-        self.assertEqual(ainfo[0].output_name, 'Preprocessed FASTA')
-        self.assertEqual(ainfo[0].files, [
-            (join(out_dir, '1.SKB8.640193.fasta'), 'preprocessed_fasta'),
-            (join(out_dir, '1.SKD8.640184.fasta'), 'preprocessed_fasta')])
-
+        prefix_to_name = df.set_index('run_prefix')['sample_name'].to_dict()
+
+        main_qsub_fp, finish_qsub_fp = spades_to_array(
+            directory, out_dir, prefix_to_name, 'http://mylink',
+            'qiita_job_id', params)
+        with open(main_qsub_fp) as fp:
+            obs_main_qsub_fp = fp.readlines()
+        with open(finish_qsub_fp) as fp:
+            obs_finish_qsub_fp = fp.readlines()
+        params['out_dir'] = out_dir
+        params['environment'] = self.environment
+        self.assertEqual(
+            ''.join(obs_main_qsub_fp), EXP_MAIN_FLASH.format(**params))
+        self.assertEqual(
+            ''.join(obs_finish_qsub_fp), EXP_FINISH.format(**params))
+
+
+EXP_MAIN = """#!/bin/bash
+#PBS -M qiita.help@gmail.com
+#PBS -N qiita_job_id
+#PBS -l nodes=1:ppn=5
+#PBS -l walltime=200:00:00
+#PBS -l mem=200g
+#PBS -o {out_dir}/qiita_job_id_${{PBS_ARRAYID}}.log
+#PBS -e {out_dir}/qiita_job_id_${{PBS_ARRAYID}}.err
+#PBS -t 1-2%8
+#PBS -l epilogue=/home/qiita/qiita-epilogue.sh
+cd {out_dir}
+{environment}
+OUTDIR={out_dir}/
+date
+hostname
+echo ${{PBS_JOBID}} ${{PBS_ARRAYID}}
```

Comment:
Just curious, why is it expected to see doubled curly braces?

Reply:
The way to scape single curly braces is to use double. Basically, `${{PBS_JOBID}}`, will  be translated to `${PBS_JOBID}` during the format action.

Reply:
ohhhhh i see that now, thanks